### PR TITLE
feat(contracts): BMT File

### PR DESCRIPTION
**Pull request by molekilla**
https://github.com/fairDataSociety/fdp-contracts/issues/17
Screenshot from 2022-05-12 15-34-00

### CODE REVIEW

**Comment by molekilla** [comment]
Ready for review
Screenshot from 2022-05-13 09-08-47

File: contracts/BMTFile.sol [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+contract BMTFile is BMTChunk {
+
+// max segment count
+  uint256 public constant MAX_SEGMENT_COUNT = 128;
+
+  // chunk bmt levels
+  uint256 public constant CHUNK_BMT_LEVELS = 7;
+
+  struct ChunkInclusionProof{
+    bytes span;
```
Comment (nugaon) [inline contracts/BMTFile.sol:16]: this should be uint64

File: test/BMTFile.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
+import {
+  Chunk,
+  fileInclusionProofBottomUp,
+  getSpanValue,
+  makeChunkedFile,
+  Utils,
+} from '@fairdatasociety/bmt-js'
+import { expect } from 'chai'
+import { ethers } from 'hardhat'
+import { BMTFile } from '../typechain'
```
Comment (nugaon) [inline test/BMTFile.ts:10]: I have problem again with the typechain import. have you rebased this branch to the master well?

File: test/BMTFile.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
+import {
```
Comment (nugaon) [inline test/BMTFile.ts:1]: bad formatting according to `prettier`

File: test/BMTFile.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  makeChunkedFile,
+  Utils,
+} from '@fairdatasociety/bmt-js'
+import { expect } from 'chai'
+import { ethers } from 'hardhat'
+import { BMTFile } from '../typechain'
+import { BMTChunkInclusionProof } from '../js-library/src/model/bmt.model'
+import FS from 'fs'
+import path from 'path'
+import {  BytesLike, hexlify } from 'ethers/lib/utils'
```
Comment (nugaon) [inline test/BMTFile.ts:14]: bad formatting by prettier.

File: test/BMTFile.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    let BMT = await ethers.getContractFactory('BMTFile')
+    bmtlib = await BMT.deploy()
+    await bmtlib.deployed()
+
+    bosBytes = Uint8Array.from(FS.readFileSync(path.join(__dirname, 'test-files', 'The-Book-of-Swarm.pdf')))
+    carrierChunkFileBytes = Uint8Array.from(FS.readFileSync(path.join(__dirname, 'test-files', 'carrier-chunk-blob')))
+  })
+
+  it('should find BMT position of the payload segment index', async () => {
+    //edge case - carrier chunk
```
Comment (nugaon) [inline test/BMTFile.ts:33]: eslint error: needs space after `//`

File: test/BMTFile.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    const chunkedFile = makeChunkedFile(fileBytes)
+    const fileHash = chunkedFile.address()
+    // segment to prove
+    const lastSegmentIndex = Math.floor((fileBytes.length - 1) / 32)
+
+    /** Gives back the file hash calculated from the inclusion proof method */
+    const testGetFileHash = (segmentIndex: number): Promise<string> => {
+      const proofChunks = fileInclusionProofBottomUp(chunkedFile, segmentIndex)
+      let proveSegment = fileBytes.slice(segmentIndex * SEGMENT_SIZE, segmentIndex * SEGMENT_SIZE + SEGMENT_SIZE)
+      //padding
```
Comment (nugaon) [inline test/BMTFile.ts:100]: same as above

File: test/BMTFile.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+
+describe('file', () => {
+  let bosBytes: Uint8Array
+  let carrierChunkFileBytes: Uint8Array
+  let bmtlib: BMTFile
+  const SEGMENT_SIZE = 32
+  const MAX_CHUNK_PAYLOAD_SIZE = 4096
+  before(async () => {
+    let BMT = await ethers.getContractFactory('BMTFile')
```
Comment (nugaon) [inline test/BMTFile.ts:24]: just use const instead of let

File: test/BMTFile.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      // check the last segment has the correct span value.
+      const fileSizeFromProof = getSpanValue(proofChunks[proofChunks.length - 1].span)
+      expect(fileSizeFromProof).eq(fileBytes.length)
+
+      const chunks = proofChunks.map(chunk => {
+        return {
+          span: chunk.span,
+          sisterSegments: chunk.sisterSegments.map(i => i as BytesLike),
+        } as BMTChunkInclusionProof
+      })
```
Comment (nugaon) [inline test/BMTFile.ts:77]: this part is really weird, why do you do this? with the typings of `proofChunks` everything works fine.

File: test/BMTFile.ts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      //padding
+      proveSegment = new Uint8Array([...proveSegment, ...new Uint8Array(SEGMENT_SIZE - proveSegment.length)])
+
+      // check the last segment has the correct span value.
+      const fileSizeFromProof = getSpanValue(proofChunks[proofChunks.length - 1].span)
+      expect(fileSizeFromProof).eq(fileBytes.length)
+
+      const chunks = proofChunks.map(chunk => {
+        return {
+          // span: getSpanValue(chunk.span),
```
Comment (nugaon) [inline test/BMTFile.ts:159]: no commented out code

**Review by nugaon — COMMENTED** [review]
LGTM but please check out my suggestions : )

File: test/BMTFile.ts (hunk 2 above)
Comment (nugaon) [reply test/BMTFile.ts:4]: ah, now I needed `npx hardhat compile` and it is fine

File: contracts/BMTFile.sol [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    bytes32[] sisterSegments;
+  }
+
+  /* 
+  * Gives back the file address that is calculated with only the inclusion proof segments
+  * and the corresponding proved segment and its position.
+  * @param _proveChunks Sister segments that will be hashed together with the calculated hashes
+  * @param _proveSegment The segment that is wanted to be validated it is subsumed under the file address
+  * @param _proveSegmentIndex the `proveSegment`'s segment index on its BMT level
+  * @param _fileLength File length
```
Comment (nugaon) [inline contracts/BMTFile.sol:26]:
```suggestion
  * @param _fileLength File length in bytes
```

File: contracts/BMTFile.sol [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  * @return _calculatedHash File address
+  */
+function fileAddressFromInclusionProof(
+  ChunkInclusionProof[] memory _proveChunks,
+  bytes32 _proveSegment,
+  uint256 _proveSegmentIndex,
+  uint256 _fileLength
```
Comment (nugaon) [inline contracts/BMTFile.sol:33]: this should be uint64

File: contracts/BMTFile.sol [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+) public pure returns (bytes32 _calculatedHash) {
+   _calculatedHash = _proveSegment;
+
+  for (uint256 i = 0; i < _proveChunks.length; i++) {
+    (uint256 parentChunkIndex, uint256 level) = getBmtIndexOfSegment(
+      _proveSegmentIndex,
+      _fileLength
```
Comment (nugaon) [inline contracts/BMTFile.sol:40]: here you ask for the `_lastChunkIndex` and you pass the `_fileLength` instead..

consider these snippet so that you can rid of this additional `_fileLength` argument, as it is already given in the `_proveChunks` array:

https://github.com/fairDataSociety/bmt-js/blob/1ca345af958d49d95359356c1881950b8f9d7838/src/file.ts#L154-L155

**Review by nugaon — CHANGES_REQUESTED** [review]

File: contracts/BMTFile.sol (hunk 12 above)
Comment (nugaon) [reply contracts/BMTFile.sol:40]: not necessary to do that in the same way, you can shift the bits as well instead

**Comment by molekilla** [comment]
Screenshot from 2022-05-22 18-44-13

Hope it goes well, after several days found an equivalent library in Solidity similar to DataView and it works well! Is MIT licensed too.

File: contracts/BMTFile.sol [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  * the segment index of a payload cannot be found in the lowest level where the splitter
+  * originally created its corresponding chunk.
+  *
+  * @param _segmentIndex the segment index of the payload
+  * @param _lastChunkIndex last chunk index on the BMT level of the segment
+  * @return level and position of the chunk that contains segment index of the payload
+  */
+  function getBmtIndexOfSegment(
+      uint256  _segmentIndex,
+      uint256 _lastChunkIndex
```
[the remark this answers is not in the document]
Comment (nugaon) [reply contracts/BMTFile.sol:86]: keep it as it is, look my comment above regarding to the `_fileLength`

File: contracts/BMTFile.sol [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  // chunk bmt levels
+  uint256 public constant CHUNK_BMT_LEVELS = 7;
+
+  struct ChunkInclusionProof{
+    uint64 span;
+    bytes32[] sisterSegments;
+  }
+
+  // Uses TypedMemView (DataView in JavaScript) to handle endianness
+  function getChunkSpanLength(
```
Comment (nugaon) [inline contracts/BMTFile.sol:24]: can't we do something from the JS side to keep the parameters and the calculations in the smart contract simple?
Like convert the 8 bytes to Bigint or to number somehow to be interpreted by the EVM by default

File: contracts/BMTFile.sol [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+pragma solidity ^0.8.0;
+
+import "./BMTChunk.sol";
+
+contract BMTFile is BMTChunk {
+  // max segment count
+  uint64 public constant MAX_SEGMENT_COUNT = 128;
+
+  // chunk bmt levels
+  uint64 public constant CHUNK_BMT_LEVELS = 7;
```
Comment (nugaon) [inline contracts/BMTFile.sol:11]:
```suggestion
  uint8 public constant CHUNK_BMT_LEVELS = 7;
```

File: contracts/BMTFile.sol [hunk 16]
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./BMTChunk.sol";
+
+contract BMTFile is BMTChunk {
+  // max segment count
+  uint64 public constant MAX_SEGMENT_COUNT = 128;
```
Comment (nugaon) [inline contracts/BMTFile.sol:8]:
```suggestion
  uint16 public constant MAX_SEGMENT_COUNT = 128;
```

File: contracts/BMTFile.sol [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    bytes32[] sisterSegments;
+  }
+
+    /**
+     * @notice          Changes the endianness of a uint64.
+     * @dev             https://graphics.stanford.edu/~seander/bithacks.html#ReverseParallel
+     * @param _b        The unsigned integer to reverse
+     * @return          v - The reversed value
+     */
+    function reverseUint64(uint64 _b) internal pure returns (uint64) {
```
Comment (nugaon) [inline contracts/BMTFile.sol:24]: couldn't it be handled on JS side?

File: test/BMTFile.ts [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+    // check segment array length for carrierChunk inclusion proof
+    const proofChunks = fileInclusionProofBottomUp(chunkedFile, lastSegmentIndex)
+    expect(proofChunks.length).eq(2) // 1 level is skipped because the segment was in a carrierChunk
+
+    /** Gives back the file hash calculated from the inclusion proof method */
+    const testGetFileHash = async (segmentIndex: number): Promise<string> => {
+      const proofChunks = fileInclusionProofBottomUp(chunkedFile, segmentIndex)
+      let proveSegment = fileBytes.slice(segmentIndex * SEGMENT_SIZE, segmentIndex * SEGMENT_SIZE + SEGMENT_SIZE)
+      // padding
+      proveSegment = new Uint8Array([...proveSegment, ...new Uint8Array(SEGMENT_SIZE - proveSegment.length)])
```
Comment (nugaon) [inline test/BMTFile.ts:57]: why is it needed?

File: README.md [hunk 19]
```diff
@@ -165,3 +165,75 @@ const chunkAddress = await bmtlib.chunkAddressFromInclusionProof(
 expect(chunkAddress).equals(ethers.utils.hexlify(chunk.address())
 ```
 
+# Verifying files with BMTFiles solidity library
```
Comment (nugaon) [inline README.md:168]: please add a glossary to the document

File: hardhat.config.ts [hunk 20]
```diff
@@ -59,6 +59,9 @@ const config: HardhatUserConfig = {
   etherscan: {
     apiKey: process.env.ETHERSCAN_API_KEY,
   },
+  mocha: {
+    timeout: 400000
```
Comment (nugaon) [inline hardhat.config.ts:63]: prettier error

File: contracts/BMTFile.sol (hunk 17 above)
Comment (molekilla) [reply contracts/BMTFile.sol:24]: too complex, because I need both span (be) and getSpanValue (le)

File: test/BMTFile.ts (hunk 18 above)
Comment (molekilla) [reply test/BMTFile.ts:57]: tried to replace test 1 and 2 without padding code but left it. This is as is , taken from bmt.js.

Comment (nugaon) [reply test/BMTFile.ts:57]: aa yeah, I see now... well, I haven't commented this test code unfortunately for myself, but the padding is needed because the sliced bytearray won't result in a full segment without it.
